os/fs/mqueue/mq_check.c

```diff
@@ -0,0 +1,108 @@
+ * fs/mqueue/mq_check.c
+ *
+ *   Copyright (C) 2007, 2009, 2014 Gregory Nutt. All rights reserved.
+ *   Author: Gregory Nutt <gnutt@nuttx.org>
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions
+ * are met:
+ *
+ * 1. Redistributions of source code must retain the above copyright
+ *    notice, this list of conditions and the following disclaimer.
+ * 2. Redistributions in binary form must reproduce the above copyright
+ *    notice, this list of conditions and the following disclaimer in
+ *    the documentation and/or other materials provided with the
+ *    distribution.
+ * 3. Neither the name NuttX nor the names of its contributors may be
+ *    used to endorse or promote products derived from this software
+ *    without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
+ * FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
+ * COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
+ * INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
+ * BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS
+ * OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED
+ * AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+ * ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ *
+ ************************************************************************/
+
+/************************************************************************
+ * Included Files
+ ************************************************************************/
+
+#include <tinyara/config.h>
+
+#include <stdio.h>
+#include <mqueue.h>
+#include <sched.h>
+#include <assert.h>
+#include <errno.h>
+#include <queue.h>
+
+#include <tinyara/mqueue.h>
+#include <tinyara/sched.h>
+
+/************************************************************************
+ * Public Functions
+ ************************************************************************/
+
+/************************************************************************
+ * Name: mq_check
+ *
+ * Description:
+ *   This function checks if a message queue descriptor is present in the
+ *   calling task's group list of mq des.
+ *
+ * Parameters:
+ *   mqdes - Message queue descriptor
+ *
+ * Return Value:
+ *   OK - if mqdes is present in the calling task group list of mqdes
+ *   ERROR - if mqdes is not present in the calling task group of mqdes
+ *
+ ************************************************************************/
+
+int mq_check(mqd_t mqdes)
+{
+	int ret = -EBADF;
+    	mqd_t mqdes_ptr;
+
```

Comment:
We need to take care of interrupt handler case here. Mq can be used to send msg from irq handler to any task. In that case, mq may be opened by any task which calls the device driver. But when irq is being handled, sched_self() will return the task running at the time of irq. This task might be different than the task which opened the device driver and the mq. So, if we are in irq hanlder, then we must NOT perform this check.